### PR TITLE
Fix incorrect debruijn handling when Record calls maybe_unify_params_in

### LINKED ISSUE #12887: Error: Anomaly "Uncaught exception Evarconv.UnableToUnify(_, _)."
#### Description of the problem
```
$ more Functor.v 
Require Import Coq.Arith.Arith.
Require Import Coq.Lists.List.
Import ListNotations.
Require Import Coq.Program.Program. (* Exports functional extensionality. *)

Record Functor (F : Type -> Type) := {
  fmap : forall A B, (A -> B) -> F A -> F B;
  fmap_identity : fmap _ _ id = id;
  fmap_compose : forall f g, compose (fmap _ _ f) (fmap _ _ g) = fmap _ _ (compose f g)
}.
$ coqc Functor.v 
File "./Functor.v", line 6, characters 0-210:
Error: Anomaly "Uncaught exception Evarconv.UnableToUnify(_, _)."
Please report at http://coq.inria.fr/bugs/.
```

#### Coq Version
```
$ coqc --version
The Coq Proof Assistant, version 8.12.0 (August 2020)
compiled on Aug 18 2020 3:26:29 with OCaml 4.10.0
```

### PR DESCRIPTION
**Pull request by SkySkimmer**
Fix #12887

### CODE REVIEW

**Comment by SkySkimmer** [comment]
The error is in 2 parts:
- incorrect debruijn handling causing a nonsense unification to be tested. This causes eg `Inductive R (x:nat) := { y : bool; z : R _ }.` to fail incorrectly
- not catching the error when a sensible unification is tested and fails, this causes bad error messages

**Review by maximedenes — APPROVED** [review]
LGTM!

**Comment by maximedenes** [comment]
@coqbot merge now

**Comment by maximedenes** [comment]
@SkySkimmer which milestone should I put?

**Comment by SkySkimmer** [comment]
milestone set

**Comment by maximedenes** [comment]
@coqbot merge now

File: vernac/comInductive.mli [hunk 1]
```diff
@@ -81,8 +81,8 @@ val template_polymorphism_candidate
    monomorphic universe context that can be made parametric in its
    conclusion sort, if one is given. *)
 
-val maybe_unify_params_in : Environ.env -> Evd.evar_map -> ninds:int -> nparams:int
+val maybe_unify_params_in : Environ.env -> Evd.evar_map -> ninds:int -> nparams:int -> binders:int
```
Comment (Zimmi48) [inline vernac/comInductive.mli:84]: Is this an API that could be used by some external plugins? If yes, I cannot backport this PR as is.

Comment (ejgallego) [reply vernac/comInductive.mli:84]: Very unlikely IMHO, in fact this function was introduced in https://github.com/coq/coq/pull/11579 , and from a quick look it should disappear if some more refactoring among record and cominductive would happen.

Comment (Zimmi48) [reply vernac/comInductive.mli:84]: Thanks! It sounds reasonably safe to backport then 😃

**Pull request merged**